junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TimeoutExtension.java

```diff
@@ -0,0 +1,199 @@
+
+class TimeoutExtension implements BeforeAllCallback, BeforeEachCallback, InvocationInterceptor {
+
+	private static final ExtensionContext.Namespace NAMESPACE = ExtensionContext.Namespace.create(Timeout.class);
+	private static final String TESTABLE_METHOD_TIMEOUT_KEY = "testable_method_timeout_from_annotation";
+
+	@Override
+	public void beforeAll(ExtensionContext context) {
+		readAndStoreTimeout(context);
+	}
+
+	@Override
+	public void beforeEach(ExtensionContext context) {
+		readAndStoreTimeout(context);
+	}
+
+	@Override
+	public void interceptBeforeAllMethod(Invocation<Void> invocation,
+			ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
+		interceptLifecycleMethod(invocation, invocationContext, extensionContext,
+			TimeoutConfiguration::getDefaultBeforeAllMethodTimeout);
+	}
+
+	@Override
+	public void interceptBeforeEachMethod(Invocation<Void> invocation,
+			ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
+		interceptLifecycleMethod(invocation, invocationContext, extensionContext,
+			TimeoutConfiguration::getDefaultBeforeEachMethodTimeout);
+	}
+
+	@Override
+	public void interceptTestMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
+			ExtensionContext extensionContext) throws Throwable {
+		interceptTestableMethod(invocation, invocationContext, extensionContext,
+			TimeoutConfiguration::getDefaultTestMethodTimeout);
+	}
+
+	@Override
+	public void interceptTestTemplateMethod(Invocation<Void> invocation,
+			ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
+		interceptTestableMethod(invocation, invocationContext, extensionContext,
+			TimeoutConfiguration::getDefaultTestTemplateMethodTimeout);
+	}
+
+	@Override
+	public <T> T interceptTestFactoryMethod(Invocation<T> invocation,
+			ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
+		return interceptTestableMethod(invocation, invocationContext, extensionContext,
+			TimeoutConfiguration::getDefaultTestFactoryMethodTimeout);
+	}
+
+	@Override
+	public void interceptAfterEachMethod(Invocation<Void> invocation,
+			ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
+		interceptLifecycleMethod(invocation, invocationContext, extensionContext,
+			TimeoutConfiguration::getDefaultAfterEachMethodTimeout);
+	}
+
+	@Override
+	public void interceptAfterAllMethod(Invocation<Void> invocation,
+			ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
+		interceptLifecycleMethod(invocation, invocationContext, extensionContext,
+			TimeoutConfiguration::getDefaultAfterAllMethodTimeout);
+	}
+
+	private void readAndStoreTimeout(ExtensionContext context) {
```

Comment:
After some mental exercise (while reviewing in the browser), I understand that this is correctly supporting lookups at the class level and method level and storing the result in the correct level in the extension context hierarchy, but... I think we should add some "developer documentation" here to make that more transparent for the community and our future selves. 😉

Reply:
I've move the method closer to where it's being used and renamed it. I hope that will do. 🙂

Reply:
ok. thanks.